### Data Loading and Inspection

In [2]:
# Load dataset:
df_support <- read.csv("support2.csv", header = TRUE, sep = ",")

df_description <- read.csv("support-variables-description.csv", header = TRUE, sep = ",")


In [48]:
df_description


Variable.Name Role    Type       
1  id            ID      Integer    
2  age           Feature Continuous 
3  death         Target  Continuous 
4  sex           Feature Categorical
5  hospdead      Target  Binary     
6  slos          Other   Continuous 
7  d.time        Other   Continuous 
8  dzgroup       Feature Categorical
9  dzclass       Feature Categorical
10 num.co        Feature Continuous 
11 edu           Feature Categorical
12 income        Feature Categorical
13 scoma         Feature Continuous 
14 charges       Feature Continuous 
15 totcst        Feature Continuous 
16 totmcst       Feature Continuous 
17 avtisst       Feature Continuous 
18 race          Feature Categorical
19 sps           Feature Continuous 
20 aps           Feature Continuous 
21 surv2m        Feature Continuous 
22 surv6m        Feature Continuous 
23 hday          Feature Integer    
24 diabetes      Feature Continuous 
25 dementia      Feature Continuous 
26 ca            Feature Categorical
27 prg2m         Feature Continuous 
28 prg6m         Feature Categorical
29 dnr           Feature Categorical
30 dnrday        Feature Continuous 
31 meanbp        Feature Continuous 
32 wblc          Feature Continuous 
33 hrt           Feature Continuous 
34 resp          Feature Continuous 
35 temp          Feature Continuous 
36 pafi          Feature Continuous 
37 alb           Feature Continuous 
38 bili          Feature Continuous 
39 crea          Feature Continuous 
40 sod           Feature Continuous 
41 ph            Feature Continuous 
42 glucose       Feature Integer    
43 bun           Feature Integer    
44 urine         Feature Integer    
45 adlp          Feature Categorical
46 adls          Feature Continuous 
47 sfdm2         Target  Categorical
   Description                                                                                              
1                                                                                                           
2  Age of the patients in years                                                                             
3  Death at any time up to National Death Index (NDI) data on 31 of December of 1994.                       
4  Gender of the patient.                                                                                   
5  Death in hospital                                                                                        
6  Days from Study Entry to Discharge                                                                       
7  Days of follow-up                                                                                        
8  The patient's disease sub category                                                                       
9  The patient's disease category                                                                           
10 The number of simultaneous diseases (or comorbidities)                                                   
11 Years of education                                                                                       
12 Income of the patient.                                                                                   
13 SUPPORT day 3 Coma Score based on Glasgow scale (predicted by a model).                                  
14 Hospital charges                                                                                         
15 Total ratio of costs to charges (RCC) cost                                                               
16 Total micro cost                                                                                         
17 Average TISS score, days 3-25                                                                            
18 Race of the patient.                                                                                     
19 SUPPORT physiology score on day 3 (predicted by a model).                                                
20 APACHE III day 3 physiology score           

In [12]:
# Inspect dataset
head(df_support)


age      death sex    hospdead slos d.time dzgroup          
1 62.84998 0     male   0         5   2029   Lung Cancer      
2 60.33899 1     female 1         4      4   Cirrhosis        
3 52.74698 1     female 0        17     47   Cirrhosis        
4 42.38498 1     female 0         3    133   Lung Cancer      
5 79.88495 0     female 0        16   2029   ARF/MOSF w/Sepsis
6 93.01599 1     male   1         4      4   Coma             
  dzclass            num.co edu ... crea      sod ph       glucose bun urine
1 Cancer             0      11  ... 1.1999512 141 7.459961 NA      NA  NA   
2 COPD/CHF/Cirrhosis 2      12  ... 5.5000000 132 7.250000 NA      NA  NA   
3 COPD/CHF/Cirrhosis 2      12  ... 2.0000000 134 7.459961 NA      NA  NA   
4 Cancer             2      11  ... 0.7999268 139       NA NA      NA  NA   
5 ARF/MOSF           1      NA  ... 0.7999268 143 7.509766 NA      NA  NA   
6 Coma               1      14  ... 0.6999512 140 7.659180 NA      NA  NA   
  adlp adls sfdm2               adlsc
1 7    7    NA                  7    
2 NA   1    <2 mo. follow-up    1    
3 1    NA   <2 mo. follow-up    0    
4 NA   NA   no(M2 and SIP pres) 0    
5 NA   2    no(M2 and SIP pres) 2    
6 NA   1    <2 mo. follow-up    1

In [4]:
# Check structure of dataset
str(df_support)
dim(df_support)


'data.frame':	9105 obs. of  47 variables:
 $ age     : num  62.8 60.3 52.7 42.4 79.9 ...
 $ death   : int  0 1 1 1 0 1 1 1 1 1 ...
 $ sex     : Factor w/ 2 levels "female","male": 2 1 1 1 1 2 2 2 2 1 ...
 $ hospdead: int  0 1 0 0 0 1 0 0 0 0 ...
 $ slos    : int  5 4 17 3 16 4 9 7 12 8 ...
 $ d.time  : int  2029 4 47 133 2029 4 659 142 63 370 ...
 $ dzgroup : Factor w/ 8 levels "ARF/MOSF w/Sepsis",..: 7 3 3 7 1 5 2 2 7 4 ...
 $ dzclass : Factor w/ 4 levels "ARF/MOSF","Cancer",..: 2 4 4 2 1 3 4 4 2 2 ...
 $ num.co  : int  0 2 2 2 1 1 1 3 2 0 ...
 $ edu     : int  11 12 12 11 NA 14 14 NA 12 11 ...
 $ income  : Factor w/ 5 levels "","$11-$25k",..: 2 2 5 5 1 1 3 1 1 3 ...
 $ scoma   : int  0 44 0 0 26 55 0 26 26 0 ...
 $ charges : num  9715 34496 41094 3075 50127 ...
 $ totcst  : num  NA NA NA NA NA NA NA NA NA NA ...
 $ totmcst : num  NA NA NA NA NA NA NA NA NA NA ...
 $ avtisst : num  7 29 13 7 18.7 ...
 $ race    : Factor w/ 6 levels "","asian","black",..: 5 6 6 6 6 6 6 6 3 4 ...
 $ sps

[1] 9105   47

After Inspecting closely, We see that for many categorical variables', the datatype are not 'Factor' eg. $ sex : chr  "male" "female" etc.

### Data Cleaning and Preprocessing

In [5]:
# Identify and Group the categorical variables as catergories of interest
cat_interest <- c("death", "sex", "hospdead", "dzgroup", "dzclass", "income", "race", "diabetes", "dementia", "ca", "dnr", "adlp", "adls", "sfdm2")

# Apply unique function to check unique categories for each variable and check their structure
unique_values <- sapply(df_support[cat_interest], unique)
unique_values

str(unique_values)


$death
[1] 0 1

$sex
[1] male   female
Levels: female male

$hospdead
[1] 0 1

$dzgroup
[1] Lung Cancer       Cirrhosis         ARF/MOSF w/Sepsis Coma             
[5] CHF               Colon Cancer      COPD              MOSF w/Malig     
8 Levels: ARF/MOSF w/Sepsis CHF Cirrhosis Colon Cancer Coma ... MOSF w/Malig

$dzclass
[1] Cancer             COPD/CHF/Cirrhosis ARF/MOSF           Coma              
Levels: ARF/MOSF Cancer Coma COPD/CHF/Cirrhosis

$income
[1] $11-$25k   under $11k            $25-$50k   >$50k     
Levels:  $11-$25k $25-$50k >$50k under $11k

$race
[1] other    white    black    hispanic asian            
Levels:  asian black hispanic other white

$diabetes
[1] 0 1

$dementia
[1] 0 1

$ca
[1] metastatic no         yes       
Levels: metastatic no yes

$dnr
[1] no dnr                          dnr after sadm  dnr before sadm
Levels:  dnr after sadm dnr before sadm no dnr

$adlp
[1]  7 NA  1  0  2  3  5  6  4

$adls
[1]  7  1  0  2 NA  5  4  6  3

$sfdm2
[1]                     <2 mo. follow-up    no(M2 and SIP pres)
[4] SIP>=30             adl>=4 (>=5 if sur) Coma or Intub      
6 Levels:  <2 mo. follow-up adl>=4 (>=5 if sur) ... SIP>=30

List of 14
 $ death   : int [1:2] 0 1
 $ sex     : Factor w/ 2 levels "female","male": 2 1
 $ hospdead: int [1:2] 0 1
 $ dzgroup : Factor w/ 8 levels "ARF/MOSF w/Sepsis",..: 7 3 1 5 2 4 6 8
 $ dzclass : Factor w/ 4 levels "ARF/MOSF","Cancer",..: 2 4 1 3
 $ income  : Factor w/ 5 levels "","$11-$25k",..: 2 5 1 3 4
 $ race    : Factor w/ 6 levels "","asian","black",..: 5 6 3 4 2 1
 $ diabetes: int [1:2] 0 1
 $ dementia: int [1:2] 0 1
 $ ca      : Factor w/ 3 levels "metastatic","no",..: 1 2 3
 $ dnr     : Factor w/ 4 levels "","dnr after sadm",..: 4 1 2 3
 $ adlp    : int [1:9] 7 NA 1 0 2 3 5 6 4
 $ adls    : int [1:9] 7 1 0 2 NA 5 4 6 3
 $ sfdm2   : Factor w/ 6 levels "","<2 mo. follow-up",..: 1 2 5 6 3 4


In [6]:
# convert categorical and ordinal variables to factors  # nolint
df_support$sex <- factor(df_support$sex)

df_support$death <- factor(df_support$death)

df_support$hospdead <- factor(df_support$hospdead)

df_support$dzgroup <- factor(df_support$dzgroup)

df_support$dzclass <- factor(df_support$dzclass)

df_support$income <- factor(df_support$income, ordered = TRUE, levels = c("under $11k", "$11-$25k", "$25-$50k", ">$50k"))

df_support$race <- factor(df_support$race)

df_support$diabetes <- factor(df_support$diabetes)

df_support$dementia <- factor(df_support$dementia)

factor(c("metastatic", "no", "yes"),
    levels = c("no", "yes", "metastatic"),
    ordered = TRUE
)

df_support$adlp <- factor(df_support$adlp, ordered = TRUE, levels = c(1, 2, 3, 4, 5, 6, 7))

df_support$adls <- factor(df_support$adls, ordered = TRUE, levels = c(1, 2, 3, 4, 5, 6, 7))

df_support$sfdm2 <- factor(df_support$sfdm2, ordered = TRUE, levels = c("no(M2 and SIP pres)", "adl>=4 (>=5 if sur)", "SIP>=30", "Coma or Intub", "<2 mo. follow-up"))


[1] metastatic no         yes       
Levels: no < yes < metastatic

In [7]:
# Check structure of dataset to confirm changes
str(df_support[cat_interest])


'data.frame':	9105 obs. of  14 variables:
 $ death   : Factor w/ 2 levels "0","1": 1 2 2 2 1 2 2 2 2 2 ...
 $ sex     : Factor w/ 2 levels "female","male": 2 1 1 1 1 2 2 2 2 1 ...
 $ hospdead: Factor w/ 2 levels "0","1": 1 2 1 1 1 2 1 1 1 1 ...
 $ dzgroup : Factor w/ 8 levels "ARF/MOSF w/Sepsis",..: 7 3 3 7 1 5 2 2 7 4 ...
 $ dzclass : Factor w/ 4 levels "ARF/MOSF","Cancer",..: 2 4 4 2 1 3 4 4 2 2 ...
 $ income  : Ord.factor w/ 4 levels "under $11k"<"$11-$25k"<..: 2 2 1 1 NA NA 3 NA NA 3 ...
 $ race    : Factor w/ 6 levels "","asian","black",..: 5 6 6 6 6 6 6 6 3 4 ...
 $ diabetes: Factor w/ 2 levels "0","1": 1 1 1 1 1 1 1 2 1 1 ...
 $ dementia: Factor w/ 2 levels "0","1": 1 1 1 1 1 1 1 1 2 1 ...
 $ ca      : Factor w/ 3 levels "metastatic","no",..: 1 2 2 1 2 2 2 2 1 1 ...
 $ dnr     : Factor w/ 4 levels "","dnr after sadm",..: 4 1 4 4 4 4 4 4 2 4 ...
 $ adlp    : Ord.factor w/ 7 levels "1"<"2"<"3"<"4"<..: 7 NA 1 NA NA NA NA NA NA NA ...
 $ adls    : Ord.factor w/ 7 levels "1"<"2"<"3"<

### MISSING VALUES ANALYSIS



In [14]:
# Count the total number of complete cases (Rows with no missing values)
num_complete_cases <- sum(complete.cases(df_support))
num_complete_cases

dim(df_support)

# We have only 106 complete cases (rows with no missing values) out of 9105 rows. This is a very small number of complete cases. In order to make the dataset usable, we will have to find identify which variables have the most missing values and drop them from the dataset. 



[1] 106

[1] 9105   47

In [20]:
# Check for Missing Values in each column and sort in descending order

na_counts_col <- colSums(is.na(df_support))

na_counts_col <- sort(na_counts_col, decreasing = TRUE)

na_counts_col[1:15]


adlp    adls   urine glucose     bun totmcst     alb  income    bili    pafi 
   7490    5975    4862    4500    4352    3475    3372    2982    2601    2325 
     ph   prg2m     edu   prg6m   sfdm2 
   2284    1649    1634    1633    1400

Variables with the most missing values are:
1. adlp : 7490 (Too many missing values, and highly correlated with other 'Activities of Daily Living Index' variables) 
2. adls : 7490
3. urine : 4862
4. glucose : 4500
5. bun : 4352
6. totmcst (Total micro cost): 3475 (Highly correlated with totcst)
7. alb : 3372  etc.

To determine which variables to drop, we will create a correlation matrix and identify the variables with high correlation. 



To start, We will drop the variables with more than 50% missing values. These are:
1. adlp and 
2.  

In [32]:
# Create df with only numeric columns for creating a correlation matrix
numeric_columns <- sapply(df_support, is.numeric)
numeric_df <- df_support[numeric_columns]

head(numeric_df)


age      slos d.time num.co edu scoma charges totcst totmcst avtisst  ...
1 62.84998  5   2029   0      11   0     9715   NA     NA       7.00000 ...
2 60.33899  4      4   2      12  44    34496   NA     NA      29.00000 ...
3 52.74698 17     47   2      12   0    41094   NA     NA      13.00000 ...
4 42.38498  3    133   2      11   0     3075   NA     NA       7.00000 ...
5 79.88495 16   2029   1      NA  26    50127   NA     NA      18.66666 ...
6 93.01599  4      4   1      14  55     6884   NA     NA       5.00000 ...
  pafi     alb      bili      crea      sod ph       glucose bun urine adlsc
1 388.0000 1.799805 0.1999817 1.1999512 141 7.459961 NA      NA  NA    7    
2  98.0000       NA        NA 5.5000000 132 7.250000 NA      NA  NA    1    
3 231.6562       NA 2.1997070 2.0000000 134 7.459961 NA      NA  NA    0    
4       NA       NA        NA 0.7999268 139       NA NA      NA  NA    0    
5 173.3125       NA        NA 0.7999268 143 7.509766 NA      NA  NA    2    
6 266.6250       NA        NA 0.6999512 140 7.659180 NA      NA  NA    1

In [49]:
# Create a corrplot to visualize the correlation between the variables
# install.packages("corrplot")
library(corrplot)
library(reshape2)

# Assuming df is your data frame
correlation_matrix <- cor(numeric_df, use = "complete.obs")

# Flatten and filter
corr_melted <- melt(correlation_matrix)

# Remove self-correlations and duplicates
corr_melted <- subset(corr_melted, Var1 != Var2)
corr_melted <- corr_melted[!duplicated(t(apply(corr_melted[, 1:2], 1, sort))), ]

# Sort by absolute correlation value
corr_melted <- corr_melted[order(-abs(corr_melted$value)), ]
top_correlations <- head(corr_melted, 100) # Top 20 correlations

top_correlations


Var1    Var2    value     
240 totmcst totcst   0.9689325
410 surv6m  surv2m   0.9603923
51  dnrday  slos     0.8952043
512 prg6m   prg2m    0.8948647
41  totcst  slos     0.8333012
206 totcst  charges  0.8328858
42  totmcst slos     0.8173748
207 totmcst charges  0.8080156
249 dnrday  totcst   0.7836419
282 dnrday  totmcst  0.7706970
342 aps     sps      0.7701623
343 surv2m  sps     -0.7460156
889 bun     crea     0.7023728
344 surv6m  sps     -0.6695264
40  charges slos     0.6188178
376 surv2m  aps     -0.5969726
309 aps     avtisst  0.5939339
216 dnrday  charges  0.5926286
274 avtisst totmcst  0.5624625
213 hday    charges  0.5623091
412 prg2m   surv2m   0.5547620
178 surv2m  scoma   -0.5482460
241 avtisst totcst   0.5477100
446 prg6m   surv6m   0.5396185
308 sps     avtisst  0.5265082
445 prg2m   surv6m   0.5264234
413 prg6m   surv2m   0.5120520
377 surv6m  aps     -0.4926692
208 avtisst charges  0.4907841
83  prg6m   d.time   0.4863762
... ...     ...     ...       
378 hday    aps      0.1985603
277 surv2m  totmcst -0.1969986
480 dnrday  hday     0.1969446
682 resp    hrt      0.1952231
244 surv2m  totcst  -0.1932936
314 prg6m   avtisst -0.1929954
45  aps     slos     0.1927756
72  scoma   d.time  -0.1924999
76  avtisst d.time  -0.1921446
626 urine   meanbp   0.1918041
887 ph      crea    -0.1895204
108 totmcst num.co  -0.1887274
411 hday    surv2m  -0.1884204
13  surv2m  age     -0.1860520
107 totcst  num.co  -0.1824859
685 alb     hrt     -0.1812029
5   edu     age     -0.1807418
172 charges scoma    0.1790324
10  avtisst age     -0.1733297
381 dnrday  aps      0.1712412
356 bili    sps      0.1700504
117 dnrday  num.co  -0.1693004
561 adlsc   prg6m   -0.1686406
7   charges age     -0.1676845
528 adlsc   prg2m   -0.1600548
422 bili    surv2m  -0.1596025
212 surv6m  charges -0.1594437
18  dnrday  age     -0.1577317
132 adlsc   num.co   0.1576658
99  adlsc   d.time  -0.1565151

In [47]:
# Assuming x and y are your variables
result <- cor.test(df_support$dnrday, df_support$avtisst, method = "pearson")

# This will give you the correlation coefficient and the p-value
print(result)



	Pearson's product-moment correlation

data:  df_support$dnrday and df_support$avtisst
t = 28.659, df = 8992, p-value < 2.2e-16
alternative hypothesis: true correlation is not equal to 0
95 percent confidence interval:
 0.2702559 0.3081321
sample estimates:
      cor 
0.2893072 

